### PR TITLE
Sign up

### PR DESCRIPTION
**Pull request by fernandatoledo**
#### Trello board reference:

* [Trello Card #2](https://trello.com/c/eB0dSHxX)

---

#### Description:

* As a user I should be able to sign up so my account is created in Target MVD

---

#### Notes:

* The gender enum was created
* The devise views for registration were generated
* Validations for the gender and name were added
---

### CODE REVIEW

File: app/controllers/users/confirmations_controller.rb [hunk 1]
```diff
@@ -0,0 +1,28 @@
+class Users::ConfirmationsController < Devise::ConfirmationsController
```
Comment (gabrielrp) [inline app/controllers/users/confirmations_controller.rb]: Remove this auto generated devise controller as you aren't overriding any method

File: app/controllers/users/omniauth_callbacks_controller.rb [hunk 2]
```diff
@@ -0,0 +1,28 @@
+class Users::OmniauthCallbacksController < Devise::OmniauthCallbacksController
```
Comment (gabrielrp) [inline app/controllers/users/omniauth_callbacks_controller.rb]: Same^

File: app/controllers/users/passwords_controller.rb [hunk 3]
```diff
@@ -0,0 +1,32 @@
+class Users::PasswordsController < Devise::PasswordsController
```
Comment (gabrielrp) [inline app/controllers/users/passwords_controller.rb]: Same^

File: app/controllers/users/registrations_controller.rb [hunk 4]
```diff
@@ -0,0 +1,60 @@
+class Users::RegistrationsController < Devise::RegistrationsController
```
Comment (gabrielrp) [inline app/controllers/users/registrations_controller.rb]: Same^

File: app/controllers/users/sessions_controller.rb [hunk 5]
```diff
@@ -0,0 +1,25 @@
+class Users::SessionsController < Devise::SessionsController
```
Comment (gabrielrp) [inline app/controllers/users/sessions_controller.rb]: Same^

File: app/controllers/users/unlocks_controller.rb [hunk 6]
```diff
@@ -0,0 +1,28 @@
+class Users::UnlocksController < Devise::UnlocksController
```
Comment (gabrielrp) [inline app/controllers/users/unlocks_controller.rb]: Same ^

File: app/views/devise/registrations/new.html.erb [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+  </div>
+
+  <div class="field">
+    <%= f.label :password_confirmation %><br />
+    <%= f.password_field :password_confirmation, autocomplete: "off" %>
+  </div>
+
+  <div class="field">
+    <%= f.label :gender %><br />
+    <%= f.select :gender, User.genders.map { |k, v| [k.to_s.humanize, k] } , :prompt => "SELECT YOUR GENDER", :selected => '',  required: true %>
```
Comment (gabrielrp) [inline app/views/devise/registrations/new.html.erb]: This line is too long

File: spec/features/sign_up_spec.rb [hunk 8]
```diff
@@ -0,0 +1,62 @@
+require "rails_helper"
+
+RSpec.feature 'Sign Up', type: :feature, js: true do
+
```
Comment (gabrielrp) [inline spec/features/sign_up_spec.rb]: Remove line

File: spec/features/sign_up_spec.rb [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+  before(:each) do
+    visit new_user_registration_path
+  end
+
+  context 'Valid parameters' do
+
```
Comment (gabrielrp) [inline spec/features/sign_up_spec.rb]: Remove line

File: spec/features/sign_up_spec.rb [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+      fill_in 'Password', with: '123456'
+      fill_in 'Password confirmation', with: '123456'
+      select 'Female', from: 'Gender'
+      click_button  'Sign up'
+      expect(page).to have_current_path(root_path)
+    end
+  end
+
+  context 'Invalid parameters' do
+
```
Comment (gabrielrp) [inline spec/features/sign_up_spec.rb]: Remove line

File: spec/features/sign_up_spec.rb [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+    end
+
+    scenario 'Name was left empty' do
+      fill_in 'Email', with: 'fernanda+3@rootstrap.com'
+      fill_in 'Password', with: '123456'
+      fill_in 'Password confirmation', with: '123456'
+      select 'Female', from: 'Gender'
+      click_button  'Sign up'
+      expect(page).to have_content 'Name can\'t be blank'
+    end
```
Comment (gabrielrp) [inline spec/features/sign_up_spec.rb]: Add new line

File: spec/features/sign_up_spec.rb [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+RSpec.feature 'Sign Up', type: :feature, js: true do
+
+  before(:each) do
+    visit new_user_registration_path
+  end
+
+  context 'Valid parameters' do
+
+    scenario 'Sucessful registration' do
+      fill_in 'Name', with: 'Fernanda Toledo'
```
Comment (gabrielrp) [inline spec/features/sign_up_spec.rb]: As you're repeating the same logic but with different params, you can define a method `fill_form_and_sign_up` for example, and send the user data as params. Something like:
```
def fill_form_and_sign_up name, email, password, password_confirmation, gender
  fill_in 'Name', with: name
  fill_in 'Email', with: email
  fill_in 'Password', with: password
  fill_in 'Password confirmation', with: password_confirmation
  select gender, from: 'Gender'
  click_button  'Sign up'
end
```

File: app/views/devise/registrations/edit.html.erb [hunk 13]
```diff
@@ -0,0 +1,43 @@
+<h2>Edit <%= resource_name.to_s.humanize %></h2>
```
Comment (gabrielrp) [inline app/views/devise/registrations/edit.html.erb]: This isn't included in this feature

File: app/controllers/application_controller.rb [hunk 14]
```diff
@@ -1,3 +1,10 @@
 class ApplicationController < ActionController::Base
   protect_from_forgery with: :exception
+  before_action :configure_permitted_parameters, if: :devise_controller?
+
+  protected
+
+  def configure_permitted_parameters
+    devise_parameter_sanitizer.permit(:sign_up, keys: [:name, :gender])
```
Comment (gonzaM) [inline app/controllers/application_controller.rb]: Use `%i()` or `%i[]`. In this case: `keys: %i(name gender)` or `keys: %i[name gender]`. The Rootstrap's gudies use the second one.

File: app/views/devise/registrations/new.html.erb [hunk 15]
```diff
@@ -0,0 +1,40 @@
...
+  </div>
+
+  <div class='field'>
+    <%= f.label :password_confirmation %><br />
+    <%= f.password_field :password_confirmation, autocomplete: 'off' %>
+  </div>
+
+  <div class="field">
+    <%= f.label :gender %><br />
+    <%= f.select :gender, User.genders.map { |k, v| [k.to_s.humanize, k] },
```
Comment (gonzaM) [inline app/views/devise/registrations/new.html.erb]: Move this logic to the model (`User.genders.map { |k, v| [k.to_s.humanize, k]`)

File: app/views/devise/registrations/new.html.erb [hunk 16]
```diff
@@ -0,0 +1,40 @@
...
+    <% end %><br />
+    <%= f.password_field :password, autocomplete: 'off' %>
+  </div>
+
+  <div class='field'>
+    <%= f.label :password_confirmation %><br />
+    <%= f.password_field :password_confirmation, autocomplete: 'off' %>
+  </div>
+
+  <div class="field">
```
Comment (gonzaM) [inline app/views/devise/registrations/new.html.erb]: Single quotes

File: app/views/devise/shared/_links.html.erb [hunk 17]
```diff
@@ -0,0 +1,25 @@
+<%- if controller_name != 'sessions' %>
+  <%= link_to "Log in", new_session_path(resource_name) %><br />
```
Comment (gonzaM) [inline app/views/devise/shared/_links.html.erb]: Single quotes here and below.

File: spec/factories/users.rb [hunk 18]
```diff
@@ -0,0 +1,10 @@
+FactoryGirl.define do
+  factory :user do
+    email "fernanda@rootstrap.com"
```
Comment (gonzaM) [inline spec/factories/users.rb]: User `Faker` gem.

File: spec/features/sign_up_spec.rb [hunk 19]
```diff
@@ -0,0 +1,46 @@
...
+def fill_form_and_sign_up name, email, password, password_confirmation, gender
+  fill_in 'Name', with: name
+  fill_in 'Email', with: email
+  fill_in 'Password', with: password
+  fill_in 'Password confirmation', with: password_confirmation
+  select gender, from: 'Gender' if gender.present?
+  click_button  'Sign up'
+end
+
+RSpec.feature 'Sign Up', type: :feature, js: true do
```
Comment (gonzaM) [inline spec/features/sign_up_spec.rb]: I think `RSpec` is not necessary here.

File: spec/features/sign_up_spec.rb [hunk 20]
```diff
@@ -0,0 +1,46 @@
...
+  before(:each) do
+    visit new_user_registration_path
+  end
+
+  context 'Valid parameters' do
```
Comment (gonzaM) [inline spec/features/sign_up_spec.rb]: Start contexts with `when` or `with`. In this case: `'when valid parameters'`.

http://www.betterspecs.org/#contexts

File: spec/features/sign_up_spec.rb [hunk 21]
```diff
@@ -0,0 +1,46 @@
...
+    scenario 'Sucessful registration' do
+      fill_form_and_sign_up 'Fernanda Toledo', 'mail@asd.com', '123456', '123456', 'Female'
+      expect(page).to have_current_path(root_path)
+    end
+  end
+
+  context 'Invalid parameters' do
```
Comment (gonzaM) [inline spec/features/sign_up_spec.rb]: Same as above.

File: spec/features/sign_up_spec.rb [hunk 22]
```diff
@@ -0,0 +1,46 @@
...
+    scenario 'Passwords dont match' do
+      fill_form_and_sign_up 'Fernanda Toledo', 'mail@asd.com', '123456', '123455', 'Female'
+      expect(page).to have_content 'Password confirmation doesn\'t match Password'
+    end
+
+    scenario 'User already exist' do
+      FactoryGirl.create(:user)
```
Comment (gonzaM) [inline spec/features/sign_up_spec.rb]: `FactoryGirl` is not necessary.

File: spec/features/sign_up_spec.rb [hunk 23]
```diff
@@ -0,0 +1,46 @@
...
+    scenario 'Sucessful registration' do
+      fill_form_and_sign_up 'Fernanda Toledo', 'mail@asd.com', '123456', '123456', 'Female'
+      expect(page).to have_current_path(root_path)
+    end
+  end
+
+  context 'Invalid parameters' do
+    scenario 'Passwords dont match' do
+      fill_form_and_sign_up 'Fernanda Toledo', 'mail@asd.com', '123456', '123455', 'Female'
+      expect(page).to have_content 'Password confirmation doesn\'t match Password'
```
Comment (gonzaM) [inline spec/features/sign_up_spec.rb]: In some places, you use double quotes for string with `'` and here you use another syntax. Try to use the same syntax always.
`'Password confirmation doesn\'t match Password'` or `"Password confirmation doesn't match Password"`

File: spec/features/welcome_spec.rb [hunk 24]
```diff
@@ -0,0 +1,11 @@
+require "rails_helper"
+
+RSpec.feature "Sign Up link", type: :feature, js: true do
```
Comment (gonzaM) [inline spec/features/welcome_spec.rb]: Same as above.

File: spec/rails_helper.rb [hunk 25]
```diff
@@ -0,0 +1,57 @@
...
+  # `post` in specs under `spec/controllers`.
+  #
+  # You can disable this behaviour by removing the line below, and instead
+  # explicitly tag your specs with their type, e.g.:
+  #
+  #     RSpec.describe UsersController, :type => :controller do
+  #       # ...
+  #     end
+  #
+  # The different available types are documented in the features, such as in
```
Comment (gonzaM) [inline spec/rails_helper.rb]: Remove commented code.

File: spec/spec_helper.rb [hunk 26]
```diff
@@ -0,0 +1,148 @@
...
+  end
+
+  config.append_after(:each) do
+    DatabaseCleaner.clean
+  end
+
+  config.expect_with :rspec do |expectations|
+    # This option will default to `true` in RSpec 4. It makes the `description`
+    # and `failure_message` of custom matchers include text for helper methods
+    # defined using `chain`, e.g.:
```
Comment (gonzaM) [inline spec/spec_helper.rb]: Same as above. Remove commented code for this file too.

**Comment by gabrielrp** [comment]
Great work! :D

**Comment by DJA89** [comment]
🎉

**Pull request merged**